# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/gl/SharedSurfaceDMABUF.cpp
# Commit: dd00e2da3a0f
# Full Hash: dd00e2da3a0f11a7cc0af9e5ece21a339f4393eb
# Author: Jeff Gilbert <jgilbert@mozilla.com>
# Date: 2020-06-11 04:04:33
# Regressor Bug: 1632249
# File Overlap Count: 3
# Description:
#   Bug 1632249 - Support compositing out-of-process WebGL. r=handyman,lsalzman,nical,geckoview-reviewers,agi,imanol
#   
#   * Majorly simplity CanvasRenderer
#   * Replace GLScreenBuffer with trivial GLSwapChain
#   * Use descriptor structs so that future SharedSurface changes aren't so painful
# ==============================================================================

diff -r 69cc1de6f754 -r dd00e2da3a0f gfx/gl/SharedSurfaceDMABUF.cpp
--- a/gfx/gl/SharedSurfaceDMABUF.cpp	Wed Jun 10 21:57:19 2020 +0000
+++ b/gfx/gl/SharedSurfaceDMABUF.cpp	Wed Jun 10 22:21:02 2020 +0000
@@ -4,42 +4,41 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 #include "SharedSurfaceDMABUF.h"
+
 #include "GLContextEGL.h"
+#include "MozFramebuffer.h"
 #include "mozilla/layers/LayersSurfaces.h"  // for SurfaceDescriptor, etc
 
 namespace mozilla::gl {
 
 /*static*/
 UniquePtr<SharedSurface_DMABUF> SharedSurface_DMABUF::Create(
-    GLContext* prodGL, const GLFormats& formats, const gfx::IntSize& size,
-    bool hasAlpha) {
-  auto flags = static_cast<WaylandDMABufSurfaceFlags>(DMABUF_TEXTURE |
-                                                      DMABUF_USE_MODIFIERS);
-  if (hasAlpha) {
-    flags = static_cast<WaylandDMABufSurfaceFlags>(flags | DMABUF_ALPHA);
-  }
-
-  RefPtr<WaylandDMABufSurface> surface =
-      WaylandDMABufSurfaceRGBA::CreateDMABufSurface(size.width, size.height,
-                                                    flags);
-  if (!surface || !surface->CreateTexture(prodGL)) {
+    const SharedSurfaceDesc& desc) {
+  const auto flags = static_cast<WaylandDMABufSurfaceFlags>(
+      DMABUF_TEXTURE | DMABUF_USE_MODIFIERS | DMABUF_ALPHA);
+  const RefPtr<WaylandDMABufSurface> surface =
+      WaylandDMABufSurfaceRGBA::CreateDMABufSurface(desc.size.width,
+                                                    desc.size.height, flags);
+  if (!surface || !surface->CreateTexture(desc.gl)) {
     return nullptr;
   }
 
-  UniquePtr<SharedSurface_DMABUF> ret;
-  ret.reset(new SharedSurface_DMABUF(prodGL, size, hasAlpha, surface));
-  return ret;
+  const auto tex = surface->GetTexture();
+  auto fb = MozFramebuffer::CreateForBacking(desc.gl, desc.size, 0, false,
+                                             LOCAL_GL_TEXTURE_2D, tex);
+  if (!fb) return nullptr;
+
+  return AsUnique(new SharedSurface_DMABUF(desc, std::move(fb), surface));
 }
 
 SharedSurface_DMABUF::SharedSurface_DMABUF(
-    GLContext* gl, const gfx::IntSize& size, bool hasAlpha,
-    RefPtr<WaylandDMABufSurface> aSurface)
-    : SharedSurface(SharedSurfaceType::EGLSurfaceDMABUF,
-                    AttachmentType::GLTexture, gl, size, hasAlpha, true),
-      mSurface(aSurface) {}
+    const SharedSurfaceDesc& desc, UniquePtr<MozFramebuffer> fb,
+    const RefPtr<WaylandDMABufSurface> surface)
+    : SharedSurface(desc, std::move(fb)), mSurface(surface) {}
 
 SharedSurface_DMABUF::~SharedSurface_DMABUF() {
-  if (!mGL || !mGL->MakeCurrent()) {
+  const auto& gl = mDesc.gl;
+  if (!gl || !gl->MakeCurrent()) {
     return;
   }
   mSurface->ReleaseTextures();
@@ -47,10 +46,14 @@
 
 void SharedSurface_DMABUF::ProducerReleaseImpl() { mSurface->FenceSet(); }
 
-bool SharedSurface_DMABUF::ToSurfaceDescriptor(
-    layers::SurfaceDescriptor* const out_descriptor) {
-  MOZ_ASSERT(mSurface);
-  return mSurface->Serialize(*out_descriptor);
+Maybe<layers::SurfaceDescriptor> SharedSurface_DMABUF::ToSurfaceDescriptor() {
+  layers::SurfaceDescriptor desc;
+  if (!mSurface->Serialize(desc)) return {};
+  return Some(desc);
 }
 
+SurfaceFactory_DMABUF::SurfaceFactory_DMABUF(GLContext& gl)
+    : SurfaceFactory({&gl, SharedSurfaceType::EGLSurfaceDMABUF,
+                      layers::TextureType::WaylandDMABUF, true}) {}
+
 }  // namespace mozilla::gl